File: projects/rust/src/registry.rs
use std::{collections::HashMap, sync::Arc};

use sonic_ws_core::{
    PacketDef, PacketSchema, PacketType, SchemaLimit, SchemaType,
    enums::{EnumPackage, EnumValue},
};

use crate::{Error, Packet, Result, packet::Metadata};

/// Ordered packet table used to map tags to one-byte wire keys.
#[derive(Debug, Clone, Default)]
pub struct PacketRegistry {
    packets: Vec<Arc<Packet>>,
    tags: HashMap<String, u8>,
}

impl PacketRegistry {
    pub fn new(packets: impl IntoIterator<Item = Packet>) -> Result<Self> {
        let mut registry = Self::default();
        for packet in packets {
            registry.push(packet)?;
        }
        Ok(registry)
    }

    pub fn push(&mut self, packet: Packet) -> Result<u8> {
        if self.packets.len() >= 255 {
            return Err(Error::Schema(
                "a registry supports at most 255 packets".into(),
            ));
        }
        if self.tags.contains_key(&packet.definition.tag) {
            return Err(Error::Schema(format!(
                "duplicate packet tag \"{}\"",
                packet.definition.tag
            )));
        }
        let key = u8::try_from(self.packets.len() + 1).expect("packet count checked");
        self.tags.insert(packet.definition.tag.clone(), key);
        self.packets.push(Arc::new(packet));
        Ok(key)
    }

    pub fn len(&self) -> usize {
        self.packets.len()
    }
    pub fn is_empty(&self) -> bool {
        self.packets.is_empty()
    }
    pub fn packets(&self) -> impl ExactSizeIterator<Item = &Arc<Packet>> {
        self.packets.iter()
    }
    pub fn key(&self, tag: &str) -> Option<u8> {
        self.tags.get(tag).copied()
    }
    pub fn by_tag(&self, tag: &str) -> Option<&Arc<Packet>> {
        self.key(tag).and_then(|key| self.by_key(key))
    }
    pub fn by_key(&self, key: u8) -> Option<&Arc<Packet>> {
        key.checked_sub(1)
            .and_then(|index| self.packets.get(index as usize))
    }

    pub fn permutation_tag_flags(&self, parent: &str, flags: &[bool]) -> Result<String> {
        let values = self.permutation_values(parent)?;
        if flags.len() != values.len() {
            return Err(Error::Value(format!(
                "variant permutation requires {} boolean flags",
                values.len()
            )));
        }
        let enabled = values
            .iter()
            .zip(flags)
            .filter_map(|(value, enabled)| enabled.then_some(value.as_str()))
            .collect::<std::collections::HashSet<_>>();
        self.find_permutation_tag(parent, &enabled)
    }

    pub fn permutation_tag_map(
        &self,
        parent: &str,
        flags: &HashMap<String, bool>,
    ) -> Result<String> {
        let values = self.permutation_values(parent)?;
        if flags.len() != values.len() || flags.keys().any(|key| !values.contains(key)) {
            return Err(Error::Value(
                "variant permutation map must define every known key".into(),
            ));
        }
        let enabled = values
            .iter()
            .filter_map(|value| flags[value].then_some(value.as_str()))
            .collect::<std::collections::HashSet<_>>();
        self.find_permutation_tag(parent, &enabled)
    }

    fn permutation_values(&self, parent: &str) -> Result<&[String]> {
        self.by_tag(parent)
            .and_then(|packet| packet.group.as_ref())
            .and_then(|group| group.permutation.as_deref())
            .ok_or_else(|| {
                Error::Value(format!(
                    "packet group \"{parent}\" does not define a VariantPermutation"
                ))
            })
    }

    fn find_permutation_tag(
        &self,
        parent: &str,
        enabled: &std::collections::HashSet<&str>,
    ) -> Result<String> {
        if enabled.is_empty() {
            return Ok(parent.to_owned());
        }
        self.packets
            .iter()
            .find_map(|packet| {
                let group = packet.group.as_ref()?;
                if group.parent != parent || group.variant.is_empty() {
                    return None;
                }
                let selected = group
                    .variant
                    .split(',')
                    .collect::<std::collections::HashSet<_>>();
                (selected == *enabled).then(|| packet.definition.tag.clone())
            })
            .ok_or_else(|| Error::Value("permutation contains an opposite combination".into()))
    }

    /// Serializes the table exactly as TypeScript and Python handshake tables do.
    pub fn serialize(&self) -> Result<Vec<u8>> {
        let mut output = Vec::new();
        for packet in &self.packets {
            serialize_packet(packet, &mut output)?;
        }
        Ok(output)
    }

    pub fn deserialize(bytes: &[u8]) -> Result<Self> {
        let mut cursor = Cursor::new(bytes);
        let mut packets = Vec::new();
        while !cursor.is_empty() {
            packets.push(deserialize_packet(&mut cursor)?);
        }
        Self::new(packets)
    }
}

fn serialize_packet(packet: &Packet, out: &mut Vec<u8>) -> Result<()> {
    write_latin1(&packet.definition.tag, out)?;
    let schema = &packet.definition.schema;
    let flags = [
        schema.dont_spread,
        false,
        schema.object,
        schema.auto_flatten,
        schema.gzip_compression,
        schema.rereference,
    ]
    .into_iter()
    .enumerate()
    .fold(0_u8, |byte, (index, value)| {
        byte | (u8::from(value) << (7 - index))
    });
    out.push(flags);

    let metadata = serde_json::to_vec(&packet.metadata())?;
    write_varint(metadata.len() as u64, out);
    out.extend(metadata);
    out.push(
        u8::try_from(schema.data_batching)
            .map_err(|_| Error::Schema("batching delay must fit in one byte".into()))?,
    );
    out.push(
        u8::try_from(packet.definition.enum_data.len())
            .map_err(|_| Error::Schema("too many enum packages".into()))?,
    );
    for package in &packet.definition.enum_data {
        serialize_enum(package, out)?;
    }

    match (&schema.packet_type, &schema.data_max, &schema.data_min) {
        (SchemaType::Single(kind), SchemaLimit::Single(max), SchemaLimit::Single(min)) => {
            write_varint(*max, out);
            write_varint(*min, out);
            out.push(*kind as u8);
        }
        (SchemaType::Object(types), SchemaLimit::Object(maxes), SchemaLimit::Object(mins)) => {
            if types.len() != maxes.len() || types.len() != mins.len() || types.len() > 255 {
                return Err(Error::Schema("object schema lengths do not match".into()));
            }
            out.push(types.len() as u8);
            for max in maxes {
                write_varint(*max, out);
            }
            for min in mins {
                write_varint(*min, out);
            }
            out.extend(types.iter().map(|kind| *kind as u8));
        }
        _ => {
            return Err(Error::Schema(
                "schema type and range shapes do not match".into(),
            ));
        }
    }
    Ok(())
}

fn deserialize_packet(cursor: &mut Cursor<'_>) -> Result<Packet> {
    let tag = cursor.read_latin1()?;
    let flags = cursor.u8()?;
    let metadata_length = cursor.varint_usize()?;
    let metadata: Metadata = serde_json::from_slice(cursor.take(metadata_length)?)?;
    let batching = i32::from(cursor.u8()?);
    let enum_count = cursor.u8()?;
    let mut enums = Vec::with_capacity(enum_count as usize);
    for _ in 0..enum_count {
        enums.push(deserialize_enum(cursor)?);
    }

    let object = flags & 0x20 != 0;
    let (packet_type, data_min, data_max) = if object {
        let count = cursor.u8()? as usize;
        let maxes = (0..count)
            .map(|_| cursor.varint())
            .collect::<Result<Vec<_>>>()?;
        let mins = (0..count)
            .map(|_| cursor.varint())
            .collect::<Result<Vec<_>>>()?;
        let types = (0..count)
            .map(|_| packet_type(cursor.u8()?))
            .collect::<Result<Vec<_>>>()?;
        (
            SchemaType::Object(types),
            SchemaLimit::Object(mins),
            SchemaLimit::Object(maxes),
        )
    } else {
        let max = cursor.varint()?;
        let min = cursor.varint()?;
        (
            SchemaType::Single(packet_type(cursor.u8()?)?),
            SchemaLimit::Single(min),
            SchemaLimit::Single(max),
        )
    };

    Ok(Packet::from_parts(
        PacketDef {
            tag,
            schema: PacketSchema {
                object,
                packet_type,
                data_min,
                data_max,
                data_batching: batching,
                max_batch_size: 10,
                dont_spread: flags & 0x80 != 0,
                auto_flatten: flags & 0x10 != 0,
                gzip_compression: flags & 0x08 != 0,
                rereference: flags & 0x04 != 0,
            },
            enum_data: enums,
        },
        metadata,
    ))
}

fn serialize_enum(package: &EnumPackage, out: &mut Vec<u8>) -> Result<()> {
    write_latin1(&package.name, out)?;
    out.push(
        u8::try_from(package.values.len())
            .map_err(|_| Error::Schema("enum supports at most 255 values".into()))?,
    );
    for value in &package.values {
        let (kind, text) = match value {
            EnumValue::String(value) => (0, value.clone()),
            EnumValue::Number(value) => (1, value.to_string()),
            EnumValue::Bool(value) => (2, value.to_string()),
            EnumValue::Undefined => (3, "undefined".into()),
            EnumValue::Null => (4, "null".into()),
        };
        write_latin1(&text, out)?;
        let length_index = out.len() - text.len() - 1;
        out.insert(length_index + 1, kind);
    }
    Ok(())
}

fn deserialize_enum(cursor: &mut Cursor<'_>) -> Result<EnumPackage> {
    let name = cursor.read_latin1()?;
    let count = cursor.u8()?;
    let mut values = Vec::with_capacity(count as usize);
    for _ in 0..count {
        let length = cursor.u8()? as usize;
        let kind = cursor.u8()?;
        let text = latin1(cursor.take(length)?);
        values.push(match kind {
            0 => EnumValue::String(text),
            1 => EnumValue::Number(
                text.parse()
                    .map_err(|_| Error::Protocol("invalid enum number".into()))?,
            ),
            2 => EnumValue::Bool(text == "true"),
            3 => EnumValue::Undefined,
            4 => EnumValue::Null,
            _ => return Err(Error::Protocol("invalid enum value type".into())),
        });
    }
    Ok(EnumPackage { name, values })
}

fn write_latin1(value: &str, out: &mut Vec<u8>) -> Result<()> {
    let bytes = value
        .chars()
        .map(|character| {
            u8::try_from(u32::from(character))
                .map_err(|_| Error::Schema("schema text must be Latin-1".into()))
        })
        .collect::<Result<Vec<_>>>()?;
    out.push(
        u8::try_from(bytes.len())
            .map_err(|_| Error::Schema("schema text exceeds 255 bytes".into()))?,
    );
    out.extend(bytes);
    Ok(())
}

fn latin1(bytes: &[u8]) -> String {
    bytes.iter().map(|byte| char::from(*byte)).collect()
}
fn write_varint(mut value: u64, output: &mut Vec<u8>) {
    loop {
        let mut byte = (value & 0x7f) as u8;
        value >>= 7;
        if value != 0 {
            byte |= 0x80;
        }
        output.push(byte);
        if value == 0 {
            break;
        }
    }
}
fn packet_type(value: u8) -> Result<PacketType> {
    Ok(match value {
        0 => PacketType::None,
        1 => PacketType::Raw,
        2 => PacketType::StringsAscii,
        3 => PacketType::StringsUtf16,
        4 => PacketType::Enums,
        5 => PacketType::Bytes,
        6 => PacketType::UBytes,
        7 => PacketType::Shorts,
        8 => PacketType::UShorts,
        9 => PacketType::VarInt,
        10 => PacketType::UVarInt,
        11 => PacketType::Deltas,
        12 => PacketType::Floats,
        13 => PacketType::Doubles,
        14 => PacketType::Booleans,
        16 => PacketType::Reserved16,
        17 => PacketType::Hex,
        _ => return Err(Error::Protocol(format!("unknown packet type {value}"))),
    })
}

struct Cursor<'a> {
    bytes: &'a [u8],
    offset: usize,
}
impl<'a> Cursor<'a> {
    fn new(bytes: &'a [u8]) -> Self {
        Self { bytes, offset: 0 }
    }
    fn is_empty(&self) -> bool {
        self.offset == self.bytes.len()
    }
    fn take(&mut self, length: usize) -> Result<&'a [u8]> {
        let end = self
            .offset
            .checked_add(length)
            .ok_or_else(|| Error::Protocol("frame length overflow".into()))?;
        if end > self.bytes.len() {
            return Err(Error::Protocol("truncated schema table".into()));
        }
        let value = &self.bytes[self.offset..end];
        self.offset = end;
        Ok(value)
    }
    fn u8(&mut self) -> Result<u8> {
        Ok(self.take(1)?[0])
    }
    fn varint(&mut self) -> Result<u64> {
        let mut value = 0_u64;
        for shift in (0..=63).step_by(7) {
            let byte = self.u8()?;
            value |= u64::from(byte & 0x7f) << shift;
            if byte & 0x80 == 0 {
                return Ok(value);
            }
        }
        Err(Error::Protocol("variable integer is too long".into()))
    }
    fn varint_usize(&mut self) -> Result<usize> {
        usize::try_from(self.varint()?)
            .map_err(|_| Error::Protocol("length exceeds this platform".into()))
    }
    fn read_latin1(&mut self) -> Result<String> {
        let length = self.u8()? as usize;
        Ok(latin1(self.take(length)?))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Packet;
    use sonic_ws_core::PacketType;

    #[test]
    fn registry_roundtrips_negotiated_metadata() {
        let packet = Packet::builder("movement.move", PacketType::VarInt)
            .data_range(3, 3)
            .schema(["dx", "dy", "dz"])
            .quantized(1000.0)
            .value_range(Some(-10.0), Some(10.0))
            .replay(true)
            .build()
            .unwrap();
        let registry = PacketRegistry::new([packet]).unwrap();
        let decoded = PacketRegistry::deserialize(&registry.serialize().unwrap()).unwrap();
        let packet = decoded.by_tag("movement.move").unwrap();
        assert_eq!(
            packet.fields.as_deref(),
            Some(["dx".into(), "dy".into(), "dz".into()].as_slice())
        );
        assert_eq!(packet.quantized.unwrap().scale, 1000.0);
        assert!(packet.replay);
    }

    #[test]
    fn registry_rejects_duplicate_tags_and_invalid_wire_types() {
        let first = Packet::builder("same", PacketType::None).build().unwrap();
        let second = Packet::builder("same", PacketType::None).build().unwrap();
        assert!(PacketRegistry::new([first, second]).is_err());
        assert!(PacketRegistry::deserialize(&[1, b'x', 0, 2, b'{', b'}', 0, 0, 0, 0, 15]).is_err());
    }
}

File: projects/go/registry.go
package sonicws

import (
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"strconv"
	"strings"
)

// Registry maps ordered packet definitions to one-byte wire keys.
type Registry struct {
	packets []*Packet
	keys    map[string]byte
}

// NewRegistry validates an ordered packet table.
func NewRegistry(packets ...*Packet) (*Registry, error) {
	if len(packets) > 254 {
		return nil, errors.New("sonicws: a registry supports at most 254 packets")
	}
	r := &Registry{packets: append([]*Packet(nil), packets...), keys: make(map[string]byte, len(packets))}
	for n, packet := range packets {
		if packet == nil {
			return nil, errors.New("sonicws: registry contains a nil packet")
		}
		if _, exists := r.keys[packet.tag]; exists {
			return nil, fmt.Errorf("sonicws: duplicate packet %q", packet.tag)
		}
		r.keys[packet.tag] = byte(n + 1)
	}
	return r, nil
}

// Len returns the number of packets.
func (r *Registry) Len() int { return len(r.packets) }

// Packet looks up a packet by tag.
func (r *Registry) Packet(tag string) (*Packet, bool) {
	key, ok := r.keys[tag]
	if !ok {
		return nil, false
	}
	return r.packets[key-1], true
}

// ByKey looks up a packet by its one-based wire key.
func (r *Registry) ByKey(key byte) (*Packet, bool) {
	if key == 0 || int(key) > len(r.packets) {
		return nil, false
	}
	return r.packets[key-1], true
}

// Key returns a packet's one-based wire key.
func (r *Registry) Key(tag string) (byte, bool) {
	key, ok := r.keys[tag]
	return key, ok
}

type packetMetadata struct {
	Schema      []string      `json:"schema,omitempty"`
	Quantized   *Quantization `json:"quantized,omitempty"`
	Min         *float64      `json:"min,omitempty"`
	Max         *float64      `json:"max,omitempty"`
	Group       *Group        `json:"group,omitempty"`
	Constructor string        `json:"constructor,omitempty"`
	Replay      bool          `json:"replay,omitempty"`
}

// MarshalBinary encodes the packet table used during negotiation.
func (r *Registry) MarshalBinary() ([]byte, error) {
	if r == nil {
		return nil, errors.New("sonicws: registry is nil")
	}
	var out []byte
	for _, packet := range r.packets {
		data, err := marshalPacket(packet)
		if err != nil {
			return nil, err
		}
		out = append(out, data...)
	}
	return out, nil
}

// ParseRegistry validates and decodes a negotiated packet table.
func ParseRegistry(data []byte) (*Registry, error) {
	c := cursor{data: data}
	var packets []*Packet
	for !c.empty() {
		if len(packets) == 254 {
			return nil, protocolf("too many packets")
		}
		packet, err := parsePacket(&c)
		if err != nil {
			return nil, err
		}
		packets = append(packets, packet)
	}
	return NewRegistry(packets...)
}

func marshalPacket(p *Packet) ([]byte, error) {
	var out []byte
	var err error
	out, err = putText(out, p.tag)
	if err != nil {
		return nil, err
	}
	flags := byte(0)
	if p.dontSpread {
		flags |= 0x80
	}
	if p.async {
		flags |= 0x40
	}
	if p.object {
		flags |= 0x20
	}
	if p.autoFlatten {
		flags |= 0x10
	}
	if p.compress {
		flags |= 0x08
	}
	if p.rereference {
		flags |= 0x04
	}
	out = append(out, flags)
	metadata, err := json.Marshal(packetMetadata{
		Schema: p.schema, Quantized: p.quantized, Min: p.valueMin, Max: p.valueMax,
		Group: p.group, Constructor: p.constructor, Replay: p.replay,
	})
	if err != nil {
		return nil, err
	}
	out = putVarint(out, uint64(len(metadata)))
	out = append(out, metadata...)
	out = append(out, p.batchMillis)
	enums := packetEnums(p)
	if len(enums) > 255 {
		return nil, errors.New("sonicws: too many packet enums")
	}
	out = append(out, byte(len(enums)))
	for _, enum := range enums {
		out, err = marshalEnum(out, enum)
		if err != nil {
			return nil, err
		}
	}
	if !p.object {
		field := p.fields[0]
		out = putVarint(out, uint64(field.Max))
		out = putVarint(out, uint64(field.Min))
		return append(out, byte(field.Type)), nil
	}
	out = append(out, byte(len(p.fields)))
	for _, field := range p.fields {
		out = putVarint(out, uint64(field.Max))
	}
	for _, field := range p.fields {
		out = putVarint(out, uint64(field.Min))
	}
	for _, field := range p.fields {
		out = append(out, byte(field.Type))
	}
	return out, nil
}

func parsePacket(c *cursor) (*Packet, error) {
	tag, err := readText(c)
	if err != nil {
		return nil, err
	}
	flags, err := c.byte()
	if err != nil {
		return nil, err
	}
	if flags&0x03 != 0 {
		return nil, protocolf("packet uses reserved flags")
	}
	metadataSize, err := c.size()
	if err != nil {
		return nil, err
	}
	metadataData, err := c.take(metadataSize)
	if err != nil {
		return nil, err
	}
	var metadata packetMetadata
	if err := json.Unmarshal(metadataData, &metadata); err != nil {
		return nil, protocolf("invalid packet metadata: %v", err)
	}
	batch, err := c.byte()
	if err != nil {
		return nil, err
	}
	enumCount, err := c.byte()
	if err != nil {
		return nil, err
	}
	enums := make([]*Enum, int(enumCount))
	for n := range enums {
		enums[n], err = parseEnum(c)
		if err != nil {
			return nil, err
		}
	}
	object := flags&0x20 != 0
	count := 1
	if object {
		value, err := c.byte()
		if err != nil {
			return nil, err
		}
		count = int(value)
		if count == 0 {
			return nil, protocolf("object packet has no fields")
		}
	}
	maxes := make([]uint32, count)
	mins := make([]uint32, count)
	for n := range maxes {
		value, err := c.varint()
		if err != nil || value > MaxValues {
			return nil, protocolf("invalid packet maximum")
		}
		maxes[n] = uint32(value)
	}
	for n := range mins {
		value, err := c.varint()
		if err != nil || value > uint64(maxes[n]) {
			return nil, protocolf("invalid packet minimum")
		}
		mins[n] = uint32(value)
	}
	fields := make([]Field, count)
	enumIndex := 0
	for n := range fields {
		kind, err := c.byte()
		if err != nil || !validType(Type(kind)) {
			return nil, protocolf("invalid packet type %d", kind)
		}
		fields[n] = Field{Type: Type(kind), Min: mins[n], Max: maxes[n]}
		if fields[n].Type == Enums {
			if enumIndex == len(enums) {
				return nil, protocolf("missing enum definition")
			}
			fields[n].Enum = enums[enumIndex]
			enumIndex++
		}
	}
	if enumIndex != len(enums) {
		return nil, protocolf("unused enum definition")
	}
	packet, err := NewPacket(PacketConfig{
		Tag: tag, Fields: fields, Object: object, Schema: metadata.Schema,
		Quantized: metadata.Quantized, Min: metadata.Min, Max: metadata.Max,
		Group: metadata.Group, Constructor: metadata.Constructor, Replay: metadata.Replay,
		DontSpread: flags&0x80 != 0, Async: flags&0x40 != 0, AutoFlatten: flags&0x10 != 0,
		Rereference: flags&0x04 != 0, Compress: flags&0x08 != 0,
		BatchMillis: batch,
	})
	if err != nil {
		return nil, protocolf("invalid packet %q: %v", tag, err)
	}
	return packet, nil
}

func packetEnums(packet *Packet) []*Enum {
	var out []*Enum
	for _, field := range packet.fields {
		if field.Type == Enums {
			out = append(out, field.Enum)
		}
	}
	return out
}

func marshalEnum(out []byte, enum *Enum) ([]byte, error) {
	var err error
	out, err = putText(out, enum.Name)
	if err != nil {
		return nil, err
	}
	out = append(out, byte(len(enum.Values)))
	for _, value := range enum.Values {
		kind, text, err := enumText(value)
		if err != nil {
			return nil, err
		}
		encoded, err := latin1(text)
		if err != nil {
			return nil, err
		}
		out = append(out, byte(len(encoded)), kind)
		out = append(out, encoded...)
	}
	return out, nil
}

func parseEnum(c *cursor) (*Enum, error) {
	name, err := readText(c)
	if err != nil {
		return nil, err
	}
	count, err := c.byte()
	if err != nil {
		return nil, err
	}
	values := make([]any, int(count))
	for n := range values {
		size, err := c.byte()
		if err != nil {
			return nil, err
		}
		kind, err := c.byte()
		if err != nil {
			return nil, err
		}
		data, err := c.take(int(size))
		if err != nil {
			return nil, err
		}
		text := latin1String(data)
		switch kind {
		case 0:
			values[n] = text
		case 1:
			if strings.ContainsAny(text, ".eE") {
				value, err := strconv.ParseFloat(text, 64)
				if err != nil || math.IsNaN(value) || math.IsInf(value, 0) {
					return nil, protocolf("invalid enum number")
				}
				values[n] = value
			} else {
				value, err := strconv.ParseInt(text, 10, 64)
				if err != nil {
					return nil, protocolf("invalid enum number")
				}
				values[n] = value
			}
		case 2:
			if text != "true" && text != "false" {
				return nil, protocolf("invalid enum boolean")
			}
			values[n] = text == "true"
		case 3:
			values[n] = Undefined{}
		case 4:
			values[n] = nil
		default:
			return nil, protocolf("invalid enum value type %d", kind)
		}
	}
	enum := &Enum{Name: name, Values: values}
	if err := validateEnum(enum); err != nil {
		return nil, protocolf("invalid enum: %v", err)
	}
	return enum, nil
}

func enumText(value any) (byte, string, error) {
	switch value := value.(type) {
	case string:
		return 0, value, nil
	case int:
		return 1, strconv.Itoa(value), nil
	case int64:
		return 1, strconv.FormatInt(value, 10), nil
	case float64:
		if math.IsNaN(value) || math.IsInf(value, 0) {
			return 0, "", errors.New("sonicws: enum number must be finite")
		}
		return 1, strconv.FormatFloat(value, 'g', -1, 64), nil
	case bool:
		return 2, strconv.FormatBool(value), nil
	case Undefined:
		return 3, "undefined", nil
	case nil:
		return 4, "null", nil
	default:
		return 0, "", fmt.Errorf("sonicws: unsupported enum value %T", value)
	}
}

// CreateManifest encodes both packet directions for offline use.
func CreateManifest(client, server *Registry) ([]byte, error) {
	if client == nil || server == nil {
		return nil, errors.New("sonicws: manifest registries cannot be nil")
	}
	clients, err := client.MarshalBinary()
	if err != nil {
		return nil, err
	}
	servers, err := server.MarshalBinary()
	if err != nil {
		return nil, err
	}
	out := append([]byte("SWSM"), Version)
	out = putVarint(out, uint64(len(clients)))
	out = append(out, clients...)
	return append(out, servers...), nil
}

// ParseManifest decodes an offline packet manifest.
func ParseManifest(data []byte) (*Registry, *Registry, error) {
	if len(data) < 5 || string(data[:4]) != "SWSM" || data[4] != Version {
		return nil, nil, protocolf("invalid packet manifest")
	}
	c := cursor{data: data[5:]}
	size, err := c.size()
	if err != nil {
		return nil, nil, err
	}
	clients, err := c.take(size)
	if err != nil {
		return nil, nil, err
	}
	client, err := ParseRegistry(clients)
	if err != nil {
		return nil, nil, err
	}
	server, err := ParseRegistry(c.rest())
	return client, server, err
}
